File: training-data/perl/priority_queue_heap.pl
use strict;
use warnings;

package MinHeap;

sub new {
    my $class = shift;
    return bless { data => [] }, $class;
}

sub insert {
    my ($self, $value) = @_;
    push @{$self->{data}}, $value;
    my $i = $#{$self->{data}};
    while ($i > 0) {
        my $parent = int(($i - 1) / 2);
        last if $self->{data}[$parent] <= $self->{data}[$i];
        @{$self->{data}}[$parent, $i] = @{$self->{data}}[$i, $parent];
        $i = $parent;
    }
}

sub extract_min {
    my $self = shift;
    my $data = $self->{data};
    return undef unless @$data;
    my $min = $data->[0];
    my $last = pop @$data;
    if (@$data) {
        $data->[0] = $last;
        my $i = 0;
        while (1) {
            my ($left, $right) = (2 * $i + 1, 2 * $i + 2);
            my $smallest = $i;
            $smallest = $left if $left <= $#$data && $data->[$left] < $data->[$smallest];
            $smallest = $right if $right <= $#$data && $data->[$right] < $data->[$smallest];
            last if $smallest == $i;
            @{$data}[$i, $smallest] = @{$data}[$smallest, $i];
            $i = $smallest;
        }
    }
    return $min;
}

package main;

my $pq = MinHeap->new();
$pq->insert($_) for (5, 3, 8, 1, 9, 2);

my @sorted;
while (defined(my $val = $pq->extract_min())) {
    push @sorted, $val;
}
print join(" ", @sorted), "\n";

File: training-data/pascal/priority_queue_heap.pas
program PriorityQueueHeapDemo;

const
  MaxSize = 100;

var
  heap: array[0..MaxSize-1] of Integer;
  size: Integer;

procedure Insert(value: Integer);
var
  i, parent, temp: Integer;
begin
  heap[size] := value;
  i := size;
  size := size + 1;
  while (i > 0) do
  begin
    parent := (i - 1) div 2;
    if heap[parent] <= heap[i] then Break;
    temp := heap[parent];
    heap[parent] := heap[i];
    heap[i] := temp;
    i := parent;
  end;
end;

function ExtractMin: Integer;
var
  i, left, right, smallest, temp: Integer;
begin
  ExtractMin := heap[0];
  size := size - 1;
  heap[0] := heap[size];
  i := 0;
  while True do
  begin
    left := 2 * i + 1;
    right := 2 * i + 2;
    smallest := i;
    if (left < size) and (heap[left] < heap[smallest]) then smallest := left;
    if (right < size) and (heap[right] < heap[smallest]) then smallest := right;
    if smallest = i then Break;
    temp := heap[i];
    heap[i] := heap[smallest];
    heap[smallest] := temp;
    i := smallest;
  end;
end;

var
  values: array[0..4] of Integer = (5, 1, 9, 3, 7);
  i: Integer;
begin
  size := 0;
  for i := 0 to 4 do Insert(values[i]);
  for i := 0 to 4 do Write(ExtractMin, ' ');
  WriteLn;
end.
